akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1104,3 +1072,44 @@ private[stream] final class Reduce[T](f: (T, T) ⇒ T) extends SimpleLinearGraph
   }
   override def toString = "Reduce"
 }
+
+/**
+ * INTERNAL API
+ */
+private[stream] final class StatefulMapConcat[In, Out](f: () ⇒ In ⇒ immutable.Iterable[Out]) extends GraphStage[FlowShape[In, Out]] {
+  val in = Inlet[In]("StatefulMapConcat.in")
+  val out = Outlet[Out]("StatefulMapConcat.out")
+  override val shape = FlowShape(in, out)
+  override def initialAttributes: Attributes = DefaultAttributes.statefulMapConcat
+
+  def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) {
+    var currentIterator: Iterator[Out] = Iterator.empty
+    var finished = false
+    val plainFun = f()
+    def hasNext = currentIterator.hasNext
+
+    def pushPull(): Unit =
+      if (hasNext) {
+        push(out, currentIterator.next())
+        if (!hasNext && finished) completeStage()
+      } else if (!finished)
+        pull(in)
+      else completeStage()
+
+    def onFinish(): Unit = if (!hasNext) completeStage() else finished = true
+
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = {
+        currentIterator = plainFun(grab(in)).iterator
```

Comment:
This makes me thing: we only ever need an Iterator, so we might as well change the signature to `() => In => Iterator[Out]`. This would also work better for Java, I think.